src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java

```
@@ -123,24 +131,92 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
+
+            // If we need to filter by detector state, make subsequent profile API calls to each detector
+            if (running != null || disabled != null || failed != null) {
+                List<CompletableFuture<GetAnomalyDetectorResponse>> profileFutures = new ArrayList<>();
+                for (SearchHit hit : hits) {
+                    CompletableFuture<GetAnomalyDetectorResponse> profileFuture = new CompletableFuture<>();
+                    profileFutures.add(profileFuture);
+                    ActionListener<GetAnomalyDetectorResponse> profileListener = ActionListener
+                        .<GetAnomalyDetectorResponse>wrap(profileResponse -> {
+                            profileFuture.complete(profileResponse);
+                        }, e -> {
+                            log.error("Failed to get anomaly detector profile.", e);
+                            profileFuture.completeExceptionally(e);
+                            listener.onFailure(e);
+                        });
+
+                    GetAnomalyDetectorRequest profileRequest = new GetAnomalyDetectorRequest(
+                        hit.getId(),
+                        Versions.MATCH_ANY,
+                        true,
+                        true,
+                        "",
+                        "",
+                        false,
+                        null
+                    );
+                    adClient.getDetectorProfile(profileRequest, profileListener);
+                }
+                CompletableFuture<List<GetAnomalyDetectorResponse>> listFuture = CompletableFuture
+                    .allOf(profileFutures.toArray(new CompletableFuture<?>[0]))
+                    .thenApply(v -> profileFutures.stream().map(CompletableFuture::join).collect(Collectors.toList()));
+                final List<GetAnomalyDetectorResponse> profileResponses = listFuture.join();
```

This throws some exceptions and needs to be inside a catch block, or will get silently swallowed in this lambda.

---

Added